crates/bitwarden-crypto/src/service/mod.rs

```diff
@@ -0,0 +1,197 @@
+    // clone it instead of passing references
+    key_stores: Arc<RwLock<Keys<SymmKeyRef, AsymmKeyRef>>>,
+}
+
+impl<SymmKeyRef: SymmetricKeyRef, AsymmKeyRef: AsymmetricKeyRef> std::fmt::Debug
+    for CryptoService<SymmKeyRef, AsymmKeyRef>
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CryptoService").finish()
+    }
+}
+
+pub struct Keys<SymmKeyRef: SymmetricKeyRef, AsymmKeyRef: AsymmetricKeyRef> {
+    symmetric_keys: Box<dyn KeyStore<SymmKeyRef>>,
+    asymmetric_keys: Box<dyn KeyStore<AsymmKeyRef>>,
+}
+
+impl<SymmKeyRef: SymmetricKeyRef, AsymmKeyRef: AsymmetricKeyRef>
+    CryptoService<SymmKeyRef, AsymmKeyRef>
+{
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            key_stores: Arc::new(RwLock::new(Keys {
+                symmetric_keys: create_key_store(),
+                asymmetric_keys: create_key_store(),
+            })),
+        }
+    }
+
+    pub fn clear(&self) {
+        let mut keys = self.key_stores.write().expect("RwLock is poisoned");
+        keys.symmetric_keys.clear();
+        keys.asymmetric_keys.clear();
+    }
+
+    /// Initiate an encryption/decryption context. This context will have read only access to the
+    /// global keys, and will have its own local key stores with read/write access. This
+    /// context-local store will be cleared up when the context is dropped.
+    ///
+    /// This is an advanced API, use with care. Prefer to instead use
+    /// `encrypt`/`decrypt`/`encrypt_list`/`decrypt_list` methods.
+    ///
+    /// One of the pitfalls of the current implementations is that keys stored in the context-local
+    /// store only get cleared automatically when dropped, and not between operations. This
+    /// means that if you are using the same context for multiple operations, you may want to
+    /// clear it manually between them.
+    pub fn context(&'_ self) -> CryptoServiceContext<'_, SymmKeyRef, AsymmKeyRef> {
+        CryptoServiceContext {
+            global: ReadOnlyGlobal(self.key_stores.read().expect("RwLock is poisoned")),
+            local_symmetric_keys: create_key_store(),
+            local_asymmetric_keys: create_key_store(),
+            _phantom: std::marker::PhantomData,
+        }
+    }
+
+    /// Initiate an encryption/decryption context. This context will have MUTABLE access to the
+    /// global keys, and will have its own local key stores with read/write access. This
+    /// context-local store will be cleared up when the context is dropped.
+    ///
+    /// This is an advanced API, use with care and ONLY when needing to modify the global keys.
+    ///
+    /// The same pitfalls as `context` apply here, but with the added risk of accidentally
+    /// modifying the global keys and leaving the service in an inconsistent state.
+    ///
+    /// TODO: We should work towards making this pub(crate)
```

Comment:
We need to update the types that generate these keys to insert them directly into the `KeyStore`. By that I mean types like `MasterKey`, `PinKey` etc.

Once those are updated to insert the keys directly instead of returning them, we won't need to use this from outside crypto again.